src/main/java/org/elasticsearch/index/query/BoolQueryParser.java

```diff
@@ -77,12 +77,12 @@ public Query parse(QueryParseContext parseContext) throws IOException, QueryPars
                         clauses.add(new BooleanClause(query, BooleanClause.Occur.MUST));
                     }
                 } else if ("must_not".equals(currentFieldName) || "mustNot".equals(currentFieldName)) {
-                    Query query = parseContext.parseInnerQuery();
+                    Query query = parseContext.parseInnerFilter();
                     if (query != null) {
                         clauses.add(new BooleanClause(query, BooleanClause.Occur.MUST_NOT));
                     }
                 } else if ("should".equals(currentFieldName)) {
-                    Query query = parseContext.parseInnerFilter();
+                    Query query = parseContext.parseInnerQuery();
                     if (query != null) {
                         clauses.add(new BooleanClause(query, BooleanClause.Occur.SHOULD));
                         if (parseContext.isFilter() && minimumShouldMatch == null) {
```

Comment:
Sorry, I think I get it now, looked at the previous BoolFilterParser code. However, if you set "minimumShouldMatch" to `1` here, does it matter that it can be overwritten by "minimumShouldMatch" field later? I guess it doesn't matter unless that value is not 0?